### PR TITLE
Feature mint

### LINKED ISSUE #27: Refactoring Backend
## User story
1. As a power user
2. I want / need easily understandable code
3. So that I can edit it as well

## Acceptance criteria
* easily understandable code for the backend

## DoD general criteria
* Feature has been fully implemented
* Feature has been merged into the mainline
* All acceptance criteria were met
* Product owner approved features
* All tests are passing
* Developers agreed to release

### PR DESCRIPTION
**Pull request by schlingling**
We  (@phildree & @MPSebastian ) implemented following changes in backend:

- file-binder & compiler for .sol-files
- deployer for contracts (ERC721)
- minter for NFTs
- reader for contracts
- all functions can be used for executing necessary functions for deploying/minting contracts/NFTs
--> ready for integration with frontend

We are not able to implemented test because testing depends on a external blockchain

closes #27 #3

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -20,7 +20,6 @@ pids
 __pycache__/
 
 # vscode
-.vscode/
```
Comment (lukas-rotsching) [inline .gitignore:23]: see my comment on .vscode/settings.json

**Review by tawfeeka — APPROVED** [review]
Nice

**Review by laila-rin — APPROVED** [review]

File: packages/nft-playbook/src/app/backend/compile_contract.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  // and call recursive function on those
+  const child_dependencies = retrieve_child_dependencies(
+    current_file_path,
+    current_file_content
+  );
+  child_dependencies.forEach((dependency) => {
+    // save cwd before recursion so that it can be restored after the recursion returns
+    const prev_cwd = process.cwd();
+    _rec_merge_all_solidity_sources(dependency, merged_sources);
+    process.chdir(prev_cwd);
```
Comment (lukas-rotsching) [inline packages/nft-playbook/src/app/backend/compile_contract.ts:74]: why do you do this all the time? i don't see `_rec_merge_all_solidity_sources()` actually changing the working directory

File: packages/nft-playbook/src/app/backend/compile_contract.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  });
+}
+
+// uses the parsed current_file_content to extract all dependencies e.g.'import "../node_modules/@openzeppelin/contracts/token/ERC721/ERC721.sol";'
+function retrieve_child_dependencies(current_file_path, current_file_content) {
```
Comment (lukas-rotsching) [inline packages/nft-playbook/src/app/backend/compile_contract.ts:79]: what's your naming scheme for functions? until here i thought you would prefix non exported (private) functions with _ but this doesn't seem to be the case.

**Comment by schlingling** [comment]
> from what i understood it looks quite good. But i think we should make a dev team1 and team2 meeting so you can explain the code to us in detail. Or at least explain what we need to know and understand

will do so upcoming sprint :) @lukas-rotsching

File: .vscode/settings.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
```
[the remark this answers is not in the document]
Comment (schlingling) [reply .vscode/settings.json:1]: We tried to activate linting for everyone before pushing to GH by using the workspace setting. Did not work as expected, so we can remove this folder/file/setting.

File: packages/nft-playbook/src/app/backend/compile_contract.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+import solc = require('solc');
+
+// retrieves all needed content and return compiled version of contract
+export const compile_contract = (path_to_contract_solidity) => {
+  const merged_sources = {}; // all dependencies in one sourcecode
+
+  // bind all the necesarry sourcecode in one <merged_sources> variable
+  const prev_cwd = process.cwd();
+  _rec_merge_all_solidity_sources(path_to_contract_solidity, merged_sources);
+  process.chdir(prev_cwd);
```
[the remark this answers is not in the document]
Comment (schlingling) [reply packages/nft-playbook/src/app/backend/compile_contract.ts:13]: @phildree please have a look

File: packages/nft-playbook/src/app/backend/compile_contract.ts (hunk 2 above)
Comment (schlingling) [reply packages/nft-playbook/src/app/backend/compile_contract.ts:74]: @phildree please have a look

File: packages/nft-playbook/src/app/backend/compile_contract.ts (hunk 3 above)
Comment (schlingling) [reply packages/nft-playbook/src/app/backend/compile_contract.ts:79]: We should introduce a pattern like you proposed! Will change this with next PR

**Comment by lukas-rotsching** [comment]
@phildree  @MPSebastian @tawfeeka  @laila-rin and me had a meeting and we merged backend and frontend code. The merged code is in a new branch called "cli-backend-merge". That's the branch that should be merged into master. I created a pull request #37 . Therefore, this pull request (#34) and #24 are obsolete and must not be merged into master.

**Pull request closed without merging**